lib/actions/environment-variables.ts

```diff
@@ -0,0 +1,299 @@
+/*
+Copyright 2016-2017 Balena
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { CommandDefinition } from 'capitano';
+
+import * as commandOptions from './command-options';
+import { normalizeUuidProp } from '../utils/normalization';
+import { DeviceVariable, ApplicationVariable } from 'balena-sdk';
+
+const getReservedPrefixes = async (): Promise<string[]> => {
+	const balena = (await import('balena-sdk')).fromSharedOptions();
+	const settings = await balena.settings.getAll();
+
+	const response = await balena.request.send({
+		baseUrl: settings.apiUrl,
+		url: '/config/vars',
```

Comment:
Atm this isn't part of the opensource api.
I would prefer a `--config` option in the respective `add` action.
We will have to add a way to specify service vars as well in the future, so imo it makes sense to prepare for that.
See: https://github.com/resin-io/resin-api/blob/master/src/core/routes/index.ts

We could maybe try catch 404's (and possibly filter the `Cannot GET /config/vars` body as well) so that this at least works out of the box for non-oss users, but it might make the code more complex and have different results based on the api used (sounds bad).

Reply:
Yes, this is going to be a problem. We definitely need to support it without `--config` for now though, as this will break some users workflows (I've looked in mixpanel and checked, there's people who have automation that depends on this working).

In future, I'd like to do a big breaking change that rewrite all this in a way that matches the SDK - no ids exposed, just get/set/delete by key. Even doing that without the vars config endpoint will be tricky though. We basically can't do any client-side validation or understand anything clever about env vars or anything unless this info is exposed somehow.

Is there a good reason why we can't just add this endpoint to open-source?

Reply:
I'm not sure about that.
Let's bring @dfunckt in the discussion!
Maybe all we need is a new versioned endpoint that returns the same results.

Reply:
Last resort would still be to try fetch the `/config/vars` (catching 404's) and if that's available and the name matches, infer that the user really mean to be using `--config`.

Reply:
This is now part of the open-source API: https://github.com/resin-io/open-balena-api/pull/2 :smile: